### PR TITLE
PUBDEV-5745, PUBDEV-5746 keep_cross_validation_models support for all models

### PR DESCRIPTION
**Pull request by seb-h2o**
This PR resolves 2 tickets as they are strongly connected:
- https://0xdata.atlassian.net/browse/PUBDEV-5745
- https://0xdata.atlassian.net/browse/PUBDEV-5746

- To offer the possibility to clean cv models during Grid Search, keep_cross_validation_models parameter is now exposed to all models.
- To improve support in AutoML, and ensure that CV models are cleaned progressively, the keep_cross_validation_models automl param is passed to all models as well as to Stacked Ensemble metalearner.
- Changed the defaults for keep_cv_models and keep_cv_predictions in R+Py clients, as they differed from defaults in Java/REST API. Also changed those defaults for AutoML to align with other models.

### CODE REVIEW

File: h2o-algos/src/main/java/hex/StackedEnsembleModel.java [hunk 1]
```diff
@@ -74,10 +76,40 @@ public StackedEnsembleModel(Key selfKey, StackedEnsembleParameters parms, Stacke
...
+          break;
+        case drf:
+          _metalearner_parameters = new DRFModel.DRFParameters();
+          break;
+        case deeplearning:
+          _metalearner_parameters = new DeepLearningModel.DeepLearningParameters();
+          break;
+      }
+    }
+
```
Comment (seb-h2o) [inline h2o-algos/src/main/java/hex/StackedEnsembleModel.java]: I"m thinking about moving this logic somewhere else: don't really like the idea of SEModel having dependencies on all other models. wdyt?

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java [hunk 2]
```diff
@@ -923,14 +924,17 @@ void defaultGBMs(Key<Grid> gridKey) {
     stackedEnsembleParameters._valid = (getValidationFrame() == null ? null : getValidationFrame()._key);
     stackedEnsembleParameters._keep_levelone_frame = true; //TODO Why is this true? Can be optionally turned off
     // Add cross-validation args
+    //TODO: cf. comment in Metalearner; assigning some buildSpec params directly to _metalearner_parameters
+    //  avoid having us to redefine defaults once more here, as well as the duplicate logic here and in  Metalearner
```
Comment (seb-h2o) [inline h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: @ledell, FYI: I have a branch proposing to remove all _metalearner_ params to keep onlythe more generic _metalearner_parameters, and everything seems to work fine.
also noticed that this change was planned somehow according to metalearner_algorithm page of user guide.
I can take associated PUBDEV-5281 if you're interested.

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java [hunk 3]
```diff
@@ -1055,55 +1059,50 @@ public void learn() {
...
+      // This will give another ensemble that is smaller than the original which takes all models into consideration.
+      List<Model> bestModelsOfEachType = new ArrayList();
+      Set<String> typesOfGatheredModels = new HashSet();
+
+      for (Model aModel : allModels) {
+        String type = getModelType(aModel);
+        if (typesOfGatheredModels.contains(type)) continue;
+        typesOfGatheredModels.add(type);
+        bestModelsOfEachType.add(aModel);
+      }
```
Comment (seb-h2o) [inline h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: no logic changes here outside removing the unused `Model m` on top and removing one unnecessary level of nested if-else

File: h2o-automl/src/main/java/ai/h2o/automl/AutoMLBuildSpec.java [hunk 4]
```diff
@@ -49,8 +49,8 @@ public AutoMLBuildControl() {
     public float max_after_balance_size = 5.0f;
 
     public int nfolds = 5;
-    public boolean keep_cross_validation_predictions = true;
-    public boolean keep_cross_validation_models = true;
+    public boolean keep_cross_validation_predictions = false;
+    public boolean keep_cross_validation_models = false;
```
Comment (seb-h2o) [inline h2o-automl/src/main/java/ai/h2o/automl/AutoMLBuildSpec.java]: @ledell would you agree to align the defaults here with the model properties defaults as defined in Model.java below

File: h2o-py/h2o/estimators/deeplearning.py [hunk 5]
```diff
@@ -36,20 +36,20 @@ class H2ODeepLearningEstimator(H2OEstimator):
     def __init__(self, **kwargs):
```
Comment (seb-h2o) [inline h2o-py/h2o/estimators/deeplearning.py]: all changes in estimators below are generated code... not sure why those are stored in git

File: h2o-py/tests/testdir_algos/automl/pyunit_automl_memory_cleanup.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        for i in range(total_runs - 1):
+            aml.train(y='CAPSULE', training_frame=train)
+        _, _, se = get_partitioned_model_names(aml.leaderboard)
+        se_all_models = [m for m in se if re.search(r'_AllModels_', m)]
+        se_best_of_family = [m for m in se if re.search(r'_BestOfFamily_', m)]
+        assert len(se) == len(se_all_models) + len(se_best_of_family)
+        assert len(se_all_models) == total_runs, "some StackedEnsemble_AllModels are missing"
+        # assert len(se_best_of_family) == total_runs, "some StackedEnsemble_BestOfFamily are missing"
+        assert len(se_best_of_family) == 1, "BUG! should be fixed by PUBDEV-5824"
+
```
Comment (seb-h2o) [inline h2o-py/tests/testdir_algos/automl/pyunit_automl_memory_cleanup.py]: @ledell found out this bug when writing tests, I have a simple fix for it: created PUBDEV-5824

File: h2o-py/tests/testdir_algos/automl/pyunit_automl_memory_cleanup.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        train, _, _ = prepare_data()
+        for i in range(total_runs - 1):
+            aml.train(y='CAPSULE', training_frame=train)
+        _, _, se = get_partitioned_model_names(aml.leaderboard)
+        se_all_models = [m for m in se if re.search(r'_AllModels_', m)]
+        se_best_of_family = [m for m in se if re.search(r'_BestOfFamily_', m)]
+        assert len(se) == len(se_all_models) + len(se_best_of_family)
+        assert len(se_all_models) == total_runs, "some StackedEnsemble_AllModels are missing"
+        # assert len(se_best_of_family) == total_runs, "some StackedEnsemble_BestOfFamily are missing"
+        assert len(se_best_of_family) == 1, "BUG! should be fixed by PUBDEV-5824"
```
Comment (seb-h2o) [inline h2o-py/tests/testdir_algos/automl/pyunit_automl_memory_cleanup.py]: @ledell found out this bug when writing tests, I have a simple fix for it: created PUBDEV-5824

File: h2o-algos/src/main/java/hex/StackedEnsembleModel.java (hunk 1 above)
Comment (seb-h2o) [reply h2o-algos/src/main/java/hex/StackedEnsembleModel.java]: tried to move this to Metalearner, along with all those model specific logic for better encapsulation, but this requires bigger refactoring that we probably don't want to include in this PR

File: h2o-algos/src/main/java/hex/ensemble/Metalearner.java [hunk 8]
```diff
@@ -59,13 +60,19 @@ void computeAutoMetalearner(){
...
+        }
+
         metaGLMBuilder._parms._seed = _metalearnerSeed;
         metaGLMBuilder._parms._non_negative = true;
         //metaGLMBuilder._parms._alpha = new double[] {0.0, 0.25, 0.5, 0.75, 1.0};
         metaGLMBuilder._parms._train = _levelOneTrainingFrame._key;
         metaGLMBuilder._parms._valid = (_levelOneValidationFrame == null ? null : _levelOneValidationFrame._key);
         metaGLMBuilder._parms._response_column = _model.responseColumn;
-        if (_model._parms._metalearner_fold_column == null) {
+        if (_model._parms._metalearner_fold_column == null) { //todo: KISS, those metalearner_* params should probably be passed directly as metalearner parameters
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/ensemble/Metalearner.java]: No TODOs in code please.

File: h2o-core/src/main/java/hex/Model.java [hunk 9]
```diff
@@ -2225,23 +2226,39 @@ public boolean testJavaScoring(Frame data, Frame model_predictions, EasyPredictM
...
       for (Key k : _output._cross_validation_models) {
         Model m = DKV.getGet(k);
-        if (m!=null) m.delete(); //delete all subparts
+        if (m!=null) {
+          m.delete(); //delete all subparts
+          c++;
+        }
       }
+      Log.info(c+" models were removed");
+      _output._cross_validation_models = null;
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/Model.java]: I don't think this is a good idea. You are making a local change to the output object but the update is not being propagated to DKV.

I suggest keeping output object unmodified.

File: h2o-core/src/main/java/hex/Model.java [hunk 10]
```diff
@@ -2225,23 +2226,39 @@ public boolean testJavaScoring(Frame data, Frame model_predictions, EasyPredictM
...
+        }
       }
+      Log.info(c+" predictions were removed");
     }
+    _output._cross_validation_predictions = null;
+
     if (_output._cross_validation_holdout_predictions_frame_id != null) {
       _output._cross_validation_holdout_predictions_frame_id.remove();
     }
+    _output._cross_validation_holdout_predictions_frame_id = null;
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/Model.java]: Same comments as above.

File: h2o-core/src/main/java/hex/ModelBuilder.java [hunk 11]
```diff
@@ -590,16 +590,18 @@ public void cv_mainModelScores(int N, ModelMetrics.MetricBuilder mbs[], ModelBui
 
     // Compute and put the cross-validation metrics into the main model
     Log.info("Computing "+N+"-fold cross-validation metrics.");
-    mainModel._output._cross_validation_models = new Key[N];
+    Key<M>[] cvModKeys = new Key[N];
+    mainModel._output._cross_validation_models = _parms._keep_cross_validation_models ? cvModKeys : null;
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/ModelBuilder.java]: I would move this to the end of the changes. It is better to keep the _output field null to prevent accidental changes.

File: h2o-core/src/main/java/hex/ModelBuilder.java [hunk 12]
```diff
@@ -628,8 +630,14 @@ public void cv_mainModelScores(int N, ModelMetrics.MetricBuilder mbs[], ModelBui
     }
     mainModel._output._cross_validation_metrics._description = N + "-fold cross-validation on training data (Metrics computed for combined holdout predictions)";
     Log.info(mainModel._output._cross_validation_metrics.toString());
+    mainModel._output._cross_validation_metrics_summary = makeCrossValidationSummaryTable(cvModKeys);
 
-    mainModel._output._cross_validation_metrics_summary = makeCrossValidationSummaryTable(mainModel._output._cross_validation_models);
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/ModelBuilder.java]: it would be good to use the same method, it is useful to see the logging messages (you have in the method deleteCrossValidationModels) especially in the cases when the models were implicitly removed

you can add a private and paremetrized method version of deleteCrossValidationModels to share the code

File: h2o-py/h2o/estimators/deeplearning.py (hunk 5 above)
Comment (michalkurka) [reply h2o-py/h2o/estimators/deeplearning.py]: Thanks for the heads-up

**Comment by michalkurka** [comment]
Looks good but shouldn't be merged without fixes in Model, please see comments.

Great tests!

File: h2o-algos/src/main/java/hex/ensemble/Metalearner.java (hunk 8 above)
Comment (seb-h2o) [reply h2o-algos/src/main/java/hex/ensemble/Metalearner.java]: ok, removing. creating ticket instead

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 12 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: that was my original version indeed, can't remember why I changed this: will restore this.

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 11 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: not sure I understand what you mean here: with the code above, the output field stays null if the property is disabled, following same pattern as with predictions below.

File: h2o-core/src/main/java/hex/Model.java (hunk 9 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/Model.java]: this nullify is probably not needed anymore given how output is now handled in  cv_mainModelScores below.
Although, what we were doing before was not very clean: typically when listing for CV models, the users was able to get the keys, but the models themselves had been removed from DKV. That's why I started to nullify here.

Good point about the DKV update, it's something I need to get used to: is there any best practice to avoid forgetting the update? like a transactional syntax or sth similar? e.g. key.withValue((v) -> use and auto update v)?

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 12 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: closed

File: h2o-algos/src/main/java/hex/ensemble/Metalearner.java (hunk 8 above)
Comment (seb-h2o) [reply h2o-algos/src/main/java/hex/ensemble/Metalearner.java]: closed

File: h2o-core/src/main/java/hex/Model.java (hunk 9 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/Model.java]: fixed and closed

File: h2o-core/src/main/java/hex/Model.java (hunk 10 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/Model.java]: closed

Comment (seb-h2o) [reply h2o-core/src/main/java/hex/Model.java]: reintroducing the null here, but with the update this time: this is used only by tests and AutoML currently, and nullifying only in AutoML through a parameter like `deleteCrossValidationPreds(boolean nullify)` multiplies the amount of almost unused methods.

File: h2o-core/src/main/java/hex/Model.java [hunk 13]
```diff
@@ -2223,24 +2224,56 @@ public boolean testJavaScoring(Frame data, Frame model_predictions, EasyPredictM
...
+    if (needsUpdate) {
+      try {
+        write_lock();
+        _output._cross_validation_predictions = null;
+        _output._cross_validation_holdout_predictions_frame_id = null;
+        update();
+      } finally {
+        unlock();
+      }
     }
```
Comment (seb-h2o) [inline h2o-core/src/main/java/hex/Model.java]: something in the likes of the following snippet may be useful in Lockable class:
```
<T extends Lockable<T>> void safeUpdate(Function<T> updateFn) {
    if (this.is_wlocked()) {
      updateFn.apply(this);
    } else {
      try {
        this.write_lock();
        updateFn.apply(this);
        this.update();
      } finally {
        this.unlock();
      }
    }
  }
```

File: h2o-core/src/main/java/hex/Model.java [hunk 14]
```diff
@@ -2223,24 +2224,56 @@ public boolean testJavaScoring(Frame data, Frame model_predictions, EasyPredictM
...
+    }
+    return c;
+  }
+
   public void deleteCrossValidationModels() {
     if (_output._cross_validation_models != null) {
-      for (Key k : _output._cross_validation_models) {
-        Model m = DKV.getGet(k);
-        if (m!=null) m.delete(); //delete all subparts
+      Log.info("Cleaning up CV Models for " + this._key.toString());
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/Model.java]: @seb-h2o it is okay to keep the keys in the object, it could be even argued that is the correct behavior - if someone reloads the object to memory you want to preserve the link to the model.

My previous comment was about not doing ` _output._cross_validation_models = null;` not about updating the model.

I think updating the model in this function is a bad idea, you cannot be sure what are the intentions of the caller and it the model should be updated (there can be other pending changes on the model) not just the change you are making in this functions.

It is perfectly fine to keep the references in the model. You cannot guarantee the consistency anyway - the user always can delete the CV models manually and "break" the reference.

File: h2o-core/src/main/java/hex/Model.java [hunk 15]
```diff
@@ -2223,24 +2224,56 @@ public boolean testJavaScoring(Frame data, Frame model_predictions, EasyPredictM
...
+        write_lock();
+        _output._cross_validation_models = null;
+        update();
+      } finally {
+        unlock();
       }
     }
   }
 
   public void deleteCrossValidationPreds() {
```
Comment (michalkurka) [inline h2o-core/src/main/java/hex/Model.java]: Same as above

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 11 above)
Comment (michalkurka) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: If you move line 594 to lets say 636 or later you will ensure nobody will access `mainModel._output._cross_validation_models` accidentally. This code will just be a little bit safer and prevents (artificial) scenarios like this

    mainModel._output._cross_validation_models = mainModel._output._cross_validation_models.clone();
    modify(mainModel._output._cross_validation_models);
    // code assumes cvModKeys and mainModel._output._cross_validation_models are the same but they are not

This is not real issue right now. Up to you.

Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: TBH mainModel is accessible in the entire method so I don't think moving lines around has a big impact, so I'll leave those lines where they were from the beginning.
The rest is psychology of reading :) some people could argue that by reading the initialization on the model output property early in the method, they'll be more aware that they shouldn't change it later... I think the difference is too tiny to deserve having a strong opinion.

Comment (seb-h2o) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: closed

File: h2o-core/src/main/java/hex/Model.java (hunk 14 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/Model.java]: let's have a quick chat about this so that I can close the ticket

Comment (seb-h2o) [reply h2o-core/src/main/java/hex/Model.java]: made it more explicit here to avoid misuse

File: h2o-algos/src/main/java/hex/StackedEnsembleModel.java (hunk 1 above)
Comment (navdeep-G) [reply h2o-algos/src/main/java/hex/StackedEnsembleModel.java]: I think omitting the addition of this method to SEModel is fine. Can add better encapsulation in another PR. How does that sound?

Is the main use of this in AutoML?

File: h2o-algos/src/main/java/hex/ensemble/Metalearner.java [hunk 16]
```diff
@@ -59,6 +59,12 @@ void computeAutoMetalearner(){
         //GLM Metalearner
         GLM metaGLMBuilder = ModelBuilder.make("GLM", _metalearnerJob, _metalearnerKey);
 
+        //Metalearner parameters
+        if (_hasMetalearnerParams) {
+            GLMModel.GLMParameters glmParams = (GLMModel.GLMParameters) _metalearner_parameters;
+            metaGLMBuilder._parms = glmParams;
+        }
+
```
Comment (navdeep-G) [inline h2o-algos/src/main/java/hex/ensemble/Metalearner.java]: Any reason for the above? How is it different from what is already happening?

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java (hunk 3 above)
Comment (navdeep-G) [reply h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: ok, sounds good.

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java [hunk 17]
```diff
@@ -1055,55 +1057,50 @@ public void learn() {
...
           bestModelKeys[i] = bestModelsOfEachType.get(i)._key;
 
-        Job<StackedEnsembleModel> bestEnsembleJob = stack("StackedEnsemble_BestOfFamily", bestModelKeys);
-        pollAndUpdateProgress(Stage.ModelTraining, "StackedEnsemble build using top model from each algorithm type", 50, this.job(), bestEnsembleJob, JobType.ModelBuild);
-      }
+      Job<StackedEnsembleModel> bestEnsembleJob = stack("StackedEnsemble_BestOfFamily", bestModelKeys);
+      pollAndUpdateProgress(Stage.ModelTraining, "StackedEnsemble build using top model from each algorithm type", 50, this.job(), bestEnsembleJob, JobType.ModelBuild);
     }
     userFeedback.info(Stage.Workflow, "AutoML: build done; built " + modelCount + " models");
     Log.info(userFeedback.toString("User Feedback for AutoML Run " + this._key + ":"));
```
Comment (navdeep-G) [inline h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: Is the above changes mostly cosmetic?

File: h2o-core/src/main/java/hex/Model.java (hunk 14 above)
Comment (michalkurka) [reply h2o-core/src/main/java/hex/Model.java]: TL;DR: agree with 3 - but with suggestions ;)

ad 1. Good points, in my opinion not a real issue. As I said previously arguments can be made that it actually does make sense to keep the keys in the model (to be eg. link them to to outputs in logs).
ad 2. I don't think it is "fiddling", models can be easily deleted even in Flow, the dead references are to be expected - what I am trying to say is that this is not an edge case.

And just to re-iterate my point of view and to clarify in general:
1. User builds a model, the model has CV models attached.
2. User decides to delete the CV models (by calling a function - we have to consider the Java API users!). IMHO modifying the model itself is not a good thing to do. We don't know what the user did with the models. They could have been saved for later or anything really. We should not destroy the link to the CV models or modify the model at all.

ad 3) Yes, we can handle it in AutoML. But I think there might be a better solution. Can we use flag  the `keep_cross_validation_models` in AutoML and set it to the model instances as well? The model can do the cleaning right after the main model is built when the Model object is still live and not finalized? We are trying to clean the model but we can prevent it from being "messy" in the first place.

Comment (michalkurka) [reply h2o-core/src/main/java/hex/Model.java]: If you remove the problematic changes I think we can merge the PR, I don't see an issue with the rest.

But I am really not convinced the "nullyfying" is the right thing to do. If the model was trained with CV enabled,   I would expect an evidence that it actually happened that way. The CV models don't need to exist but I would expect keys to n CV models. That is no an unreasonable invariant. If you do not want to keep the keys, tell the model to clean the CV models in the process of model building. That also makes sense.

File: h2o-algos/src/main/java/hex/ensemble/Metalearner.java (hunk 16 above)
Comment (seb-h2o) [reply h2o-algos/src/main/java/hex/ensemble/Metalearner.java]: for some reason, the _metalearner_parameters were completely ignored when computing the autoMetalearner, and I prefer to pass new params through this _metalearner_parameters object rather than adding a new _metalearner_xxx params all the way down.
Actually, maybe we can think about removing those legacy _metalearner_xxx params to rely only on _metalearner_parameters, code would benefit in simplicity and clarity, wdyt?

File: h2o-algos/src/main/java/hex/StackedEnsembleModel.java (hunk 1 above)
Comment (seb-h2o) [reply h2o-algos/src/main/java/hex/StackedEnsembleModel.java]: yes, it's mainly used in AutoML for now, but I have some idea to encapsulate all the metalearner logic in its dedicated class: will do that later and propose a PR.
will leave it as it is now if no one complains

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java (hunk 17 above)
Comment (seb-h2o) [reply h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: yes, see previous comment

File: h2o-core/src/main/java/hex/Model.java (hunk 14 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/Model.java]: this is why a quick chat is often more efficient than github :)
> Can we use flag the keep_cross_validation_models in AutoML and set it to the model instances as well?...` and the rest of this sentence.

This is exactly what I've been adding to `Model` and `ModelBuilder`. And this leads to CV models being consistently null in any case (automl or not) if the keep_cv_models param is set to false, and it behaves exactly how you suggeset in your last comment.
- Now comes the inconsistency in AutoML: the issue is not with `keep_cross_validation_models` but with `keep_cross_validation_predictions`. I believe those should behave the same way, and now they do in most cases. Except that AutoML requires to temporarily set `keep_cross_validation_predictions=true` to all models so that we can train the SE Models, and we can only clean up the predictions after the SE models are ready: this is where the `deleteCrossValidationPreds` enters the scene.

TBH, `deleteCrossValidationModels` is probably useless now, was a safety net and I believe it can be safely removed. But this is not the case for `deleteCrossValidationPreds`.

> If the model was trained with CV enabled, I would expect an evidence that it actually happened that way. The CV models don't need to exist but I would expect to see keys to the n CV models. That is not an unreasonable invariant.

I agree, this is not unreasonable to show some clue to the user that the models were computed using CV, but he has one: it's named `cross_validation_metrics`, those are always available and far more informative than dead keys ;)

To sum up:
1. I believe I can remove `deleteCrossValidationModels` completely as this one behaves consistently with this PR, just need to run a couple of tests and confirm it's safe to remove.
2. I'm not happy with `cross_validation_predictions` that behaves differently when training a single model or a grid, or AutoML. And also, it doesn't behave like `cross_validation_models`: I got convinced to "nullify" those predictions once I printed it with python to see `[ , , , , ]`. As a user, if I consume this, I don't know what to think about this empty-but-actually-not-really-empty array...

Comment (michalkurka) [reply h2o-core/src/main/java/hex/Model.java]: Okay, can you then please undo the changes in `deleteCrossValidationModels`. Remove methods:

- deleteAll
- deleteCrossValidationModels(boolean nullify)

and revert all the changes in `deleteCrossValidationPreds`?

Happy to quick chat anytime - please ping me on Slack. Github comments are too asynchronous I usually go over them when I have free time.

File: h2o-core/src/main/java/hex/Model.java (hunk 15 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/Model.java]: closed

File: h2o-core/src/main/java/hex/Model.java (hunk 14 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/hex/Model.java]: removed nullify/update and closing this

File: h2o-automl/src/main/java/ai/h2o/automl/AutoMLBuildSpec.java (hunk 4 above)
Comment (seb-h2o) [reply h2o-automl/src/main/java/ai/h2o/automl/AutoMLBuildSpec.java]: @ledell this requires documentation changes, so please let me ask confirmation again before I merge this, thanks

**Comment by ledell** [comment]
@seb-h2o I think we should definitely set `keep_cross_validation_models` to false by default.  I am on the fence about setting `keep_cross_validated_predictions` to false; that will mean the user can't create new stacked ensembles with the base models by default, though I don't know that this is a typical use-case.  Throwing away cv preds by default would be cleaner in the DKV and also mimic the default for all H2O algos (which doesn't matter too much). 

If you try to create new stacked ensembles with the AutoML models when the cv preds have been deleted, you will see an error like this:
```
water.exceptions.H2OIllegalArgumentException: Failed to find the xval predictions frame. . .  Looks like keep_cross_validation_predictions wasn't set when building the models, or the frame was deleted.

water.exceptions.H2OIllegalArgumentException: Failed to find the xval predictions frame. . .  Looks like keep_cross_validation_predictions wasn't set when building the models, or the frame was deleted.
	at hex.ensemble.StackedEnsemble$StackedEnsembleDriver.prepareTrainingLevelOneFrame(StackedEnsemble.java:166)
	at hex.ensemble.StackedEnsemble$StackedEnsembleDriver.computeImpl(StackedEnsemble.java:236)
	at hex.ModelBuilder$Driver.compute2(ModelBuilder.java:215)
	at water.H2O$H2OCountedCompleter.compute(H2O.java:1267)
	at jsr166y.CountedCompleter.exec(CountedCompleter.java:468)
	at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:263)
	at jsr166y.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:974)
	at jsr166y.ForkJoinPool.runWorker(ForkJoinPool.java:1477)
	at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104)

Error: water.exceptions.H2OIllegalArgumentException: Failed to find the xval predictions frame. . .  Looks like keep_cross_validation_predictions wasn't set when building the models, or the frame was deleted.
```

Let me think about this for a min and decide tomorrow (if you more thoughts on the defaults, let me know).

I created a JIRA for changing the default(s) here so that it's obvious in the release notes that we changed this value: https://0xdata.atlassian.net/browse/PUBDEV-5886

Are you planning on making the docs change (regarding default value) in this PR, or do you want to follow up?

**Comment by seb-h2o** [comment]
@ledell regarding the CV predictions, this is currently true, but mainly because the `keep_cross_validation_predictions` param manages both cv predictions and cv holdout predictions.
As we noticed when discussing this with Michal, SE only requires cv holdout predictions. I even created a ticket https://0xdata.atlassian.net/browse/PUBDEV-5853 if we want to change this.

Basically, part of the PUBDEV-5853 ticket could be to always keep holdout predictions in AutoML, regardless of what the user decides to do with CV predictions... or, we could also add another param.

Concerning the docs change, I can do the changes in this PR: we should just agree on the defaults first :)
